src/graph/transform/knn.cc

```diff
@@ -0,0 +1,117 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file graph/transform/knn.cc
+ * \brief k-nearest-neighbor (KNN) implementation
+ */
+
+#include "kdtree_ndarray_adaptor.h"
+#include "knn.h"
+#include <vector>
+#include <dgl/runtime/registry.h>
+#include <dgl/runtime/packed_func.h>
+#include "../../array/check.h"
+
+using namespace dgl::runtime;
+using namespace dgl::transform::knn_utils;
+namespace dgl {
+namespace transform {
+namespace impl {
+
+/*! \brief The kd-tree implementation of K-Nearest Neighbors */
+template <typename FloatType, typename IdType>
+void KdTreeKNN(const NDArray data_points, const IdArray data_offsets,
+               const NDArray query_points, const IdArray query_offsets,
+               const int k, IdArray result) {
+  int64_t batch_size = data_offsets->shape[0] - 1;
+  int64_t feature_size = data_points->shape[1];
+  IdType * data_offsets_data = static_cast<IdType*>(data_offsets->data);
+  IdType * query_offsets_data = static_cast<IdType*>(query_offsets->data);
+  IdType * result_data = static_cast<IdType*>(result->data);
+  FloatType * query_points_data = static_cast<FloatType*>(query_points->data);
+
+  for (int64_t b = 0; b < batch_size; ++b) {
+    auto d_offset = data_offsets_data[b];
+    auto d_length = data_offsets_data[b + 1] - d_offset;
+    auto q_offset = query_offsets_data[b];
+    auto q_length = query_offsets_data[b + 1] - q_offset;
+    auto out_offset = 2 * k * q_offset;
+
+    // create view for each segment
+    NDArray current_data_points = const_cast<NDArray*>(&data_points)->CreateView(
```

Comment:
Why `const_cast`?

Reply:
Because `CreateView` is not a const member function.

Reply:
I see.

As `current_data_points` are not mutated you need a `const` qualifier before.

@jermainewang I feel that we need to provide a `const NDArray NDArray::CreateView(...) const` member function.  What do you prefer?  Or you think `const_cast` is OK?